base/set.jl

```diff
@@ -191,6 +191,66 @@ function unique(f::Callable, C)
 end
 
 """
+    unique!(A::AbstractVector)
+
+Remove duplicate items as determined by [`isequal`](@ref), then return the modified `A`.
+
+```jldoctest
+julia> unique!([1, 1, 1])
+1-element Array{Int64,1}:
+ 1
+
+julia> A = [7, 3, 2, 3, 7, 5];
+
+julia> unique!(A);
+
+julia> A
+4-element Array{Int64,1}:
+ 7
+ 3
+ 2
+ 5
+```
+"""
+function unique!(A::AbstractVector)
+    if isempty(A)
+        return A
+    elseif issorted(A) || issorted(A, rev=true)
+        # If A is sorted, then we only need to keep track of one element and add that to A
+        # every time that we see a new element.
+        idxs = eachindex(A)
+        m = first(A)
+        j = start(idxs)
+        for i in idxs
+            x = A[i]
+            if x != m
+                m, j = next(idxs, j)
+                m = A[j] = x
+            end
+        end
+        count = j - start(idxs) + 1
+    else
+        # If A is not sorted, then we will need to keep track of all of the elements that
+        # we have seen so far.
+        seen = Set{eltype(A)}()
+        idxs = eachindex(A)
+        m = n = start(idxs)
+        count = 0
+        while !done(A, n)
```

Comment:
Come to think of it, this can be simplified to `for x in A` since `i` isn't needed. Then `j` can be renamed to `i`, and `m` would be clearer as `state`. You can also compute `count` after the loop as in the other branch to save one instruction per iteration.